### ISSUE TITLE
e2e kube-proxy mode detection assumes kube-proxy listens on localhost

### ISSUE THREAD

**Issue by invidian** [kubernetes/kubernetes#94337]
**What happened**:

When running the following e2e tests:
```sh
[sig-network] Services
  should have session affinity timeout work for service with type clusterIP [LinuxOnly] [Conformance]
[sig-network] Services
  should have session affinity timeout work for NodePort service [LinuxOnly] [Conformance]
```

Tests fails to detect kube-proxy mode, when kube-proxy is not listening on `localhost` address. In my configuration, `kube-proxy` only listens on node IP:
```
[sig-network] Services
  should have session affinity timeout work for NodePort service [LinuxOnly] [Conformance]
  /workspace/anago-v1.19.0-rc.4.197+594f888e19d8da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:597
[BeforeEach] [sig-network] Services
  /workspace/anago-v1.19.0-rc.4.197+594f888e19d8da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:174
STEP: Creating a kubernetes client
Aug 30 08:48:43.260: INFO: >>> kubeConfig: /tmp/kubeconfig-099743204
STEP: Building a namespace api object, basename services
STEP: Binding the e2e-test-privileged-psp PodSecurityPolicy to the default service account in services-2653
STEP: Waiting for a default service account to be provisioned in namespace
[BeforeEach] [sig-network] Services
  /workspace/anago-v1.19.0-rc.4.197+594f888e19d8da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/service.go:782
[It] should have session affinity timeout work for NodePort service [LinuxOnly] [Conformance]
  /workspace/anago-v1.19.0-rc.4.197+594f888e19d8da/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:597
STEP: creating service in namespace services-2653
Aug 30 08:48:45.463: INFO: Running '/usr/local/bin/kubectl --kubeconfig=/tmp/kubeconfig-099743204 exec --namespace=services-2653 kube-proxy-mode-detector -- /bin/sh -x -c curl -q -s --connect-timeout 1 http://localhost:10249/proxyMode'
Aug 30 08:48:45.665: INFO: rc: 7
Aug 30 08:48:45.673: INFO: Waiting for pod kube-proxy-mode-detector to disappear
Aug 30 08:48:45.677: INFO: Pod kube-proxy-mode-detector still exists
Aug 30 08:48:47.678: INFO: Waiting for pod kube-proxy-mode-detector to disappear
Aug 30 08:48:47.684: INFO: Pod kube-proxy-mode-detector still exists
Aug 30 08:48:49.678: INFO: Waiting for pod kube-proxy-mode-detector to disappear
Aug 30 08:48:49.683: INFO: Pod kube-proxy-mode-detector still exists
Aug 30 08:48:51.682: INFO: Waiting for pod kube-proxy-mode-detector to disappear
Aug 30 08:48:51.687: INFO: Pod kube-proxy-mode-detector still exists
Aug 30 08:48:53.678: INFO: Waiting for pod kube-proxy-mode-detector to disappear
Aug 30 08:48:53.681: INFO: Pod kube-proxy-mode-detector no longer exists
Aug 30 08:48:53.681: INFO: Couldn't detect KubeProxy mode - test failure may be expected: error running /usr/local/bin/kubectl --kubeconfig=/tmp/kubeconfig-099743204 exec --namespace=services-2653 kube-proxy-mode-detector -- /bin/sh -x -c curl -q -s --connect-timeout 1 http://localhost:10249/proxyMode:
Command stdout:

stderr:
+ curl -q -s --connect-timeout 1 http://localhost:10249/proxyMode
command terminated with exit code 7

error:
exit status 7
```

Because of that, following tests modifier is not applied: https://github.com/kubernetes/kubernetes/pull/89854/files, which makes tests fail.

**What you expected to happen**:

Either:
- detecting kube-proxy mode should be more robust and be able to detect proxy mode somehow
- requirement of `kube-proxy` listening on localhost of node should be documented and covered by e2e tests

**How to reproduce it (as minimally and precisely as possible)**:

- Deploy `kube-proxy` DaemonSet on the cluster using https://github.com/flexkube/kube-proxy-helm-chart version `0.2.1`. Version `0.2.2` is already patched with a workaround.
- run sonobuoy conformance tests

**Anything else we need to know?**:

Hardcoded `localhost`: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/service.go#L3689

**Environment**:
- Kubernetes version (use `kubectl version`):
  ```
  Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"archive", BuildDate:"2020-07-01T16:28:46Z", 
  GoVersion:"go1.14.4", Compiler:"gc", Platform:"linux/amd64"}
  Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:23:04Z", 
  GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
  ```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by neolit123** [comment]
/kind failing-test
/area test

possible we should support your use case.

**Comment by invidian** [comment]
BTW, with the workaround, it seems those 2 tests are still flaky to me :disappointed: Logs from the test: https://pastebin.com/F6eLN5hH

**Comment by athenabot** [comment]
/triage unresolved

Comment `/remove-triage unresolved` when the issue is assessed and confirmed.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by BenTheElder** [comment]
@aojea

**Comment by aojea** [comment]
I think that the title is misleading, it should be "e2e test kube-proxy detection assumes that kube-proxy listens on localhost", right?

kube-proxy binds only to localhost by default for security reasons 
https://github.com/kubernetes/kubernetes/pull/74300#issuecomment-510906208
more context here https://github.com/kubernetes/kubernetes/issues/74011

I assume that is the reason why the detector has a check for `localhost` hardcoded, and looking at your scripts seems that you are binding it directly you the host ip. I don't have any strong opinion on this, but seems you find a workaround, you can also bind kube-proxy to 0.0.0.0 so it listens in both localhost and the host ip.

> BTW, with the workaround, it seems those 2 tests are still flaky to me  Logs from the test: https://pastebin.com/F6eLN5hH

you can read more about IPVS and session affinity here https://github.com/kubernetes/kubernetes/issues/89358 , 
The workaround for the e2e tests was merged in 1.19 https://github.com/kubernetes/kubernetes/pull/89854/commits/fc3d282b101beaf80505d9ea7570098de5af8206 , I'm not familiar with  sonobuoy, but you have to check that the version you are running has the e2e patch included

**Comment by invidian** [comment]
>I think that the title is misleading, it should be "e2e test kube-proxy detection assumes that kube-proxy listens on localhost", right?

You're right, I fixed that.

>I assume that is the reason why the detector has a check for localhost hardcoded, and looking at your scripts seems that you are binding it directly to the host ip. I don't have any strong opinion on this, but seems you've found a workaround, another option is to bind kube-proxy to 0.0.0.0 so it listens in both localhost and the host ip.

Thanks for the insights. I somehow missed `--metrics-bind-address` flag, which I think I can use instead of making kube-proxy bind to host IP, as primary goal for that is to be able to scrape the metrics from it.

I'm not sure how to proceed from here with this issue. It would be nice to have this documented somewhere, instead of just in an issue, but since listening on localhost is the default value, it is sort of OK to make this assumption I think. But ideally that could be detected somehow by test suite and done automatically :) 

>you can read more about IPVS and session affinity here #89358 ,

Yeah, I went trough all those PRs/Issues, but I couldn't find any hints what to configure either on host or kube-proxy level to make it work. As far as I understand, it should just work :tm:, but it doesn't :confused: While running tests multiple time, always either one or the other are failing. And if I re-run the tests, they eventually pass, but this cannot be submitted as conformance results :/ 

>The workaround for the e2e tests was merged in 1.19 fc3d282 , I'm not familiar with sonobuoy, but you have to check that the version you are running has the e2e patch included

It does, as I have seen this error message: https://github.com/kubernetes/kubernetes/commit/fc3d282b101beaf80505d9ea7570098de5af8206#diff-3e87abdcaa68e4594b468bc5e5d99551R3240

**Comment by aojea** [comment]
> Yeah, I went trough all those PRs/Issues, but I couldn't find any hints what to configure either on host or kube-proxy level to make it work

there is no way to configure it, at least with current kernel versison, is a limitation in IPVS, hence the test for IPVS mode has a longer timeout. kubernetes does not put any limit on the stickiness duration, it just has to stop to be sticky after the timeout configured.

> It does, as I have seen this error message: [fc3d282#diff-3e87abdcaa68e4594b468bc5e5d99551R3240](https://github.com/kubernetes/kubernetes/commit/fc3d282b101beaf80505d9ea7570098de5af8206#diff-3e87abdcaa68e4594b468bc5e5d99551R3240)

That log was there before, seems there is a release with 1.19, are you using that?

https://github.com/vmware-tanzu/sonobuoy/releases/tag/v0.19.0

**Comment by invidian** [comment]
>That log was there before, seems there is a release with 1.19, are you using that?

Yes: https://github.com/flexkube/libflexkube/blob/v0.4.0/e2e/Dockerfile#L12.

**Comment by aojea** [comment]
@invidian the logs with the failures you linked shows that the test is not able to detect the proxy-mode, hence not applying the timeout.

**Comment by invidian** [comment]
>@invidian the logs with the failures you linked shows that the test is not able to detect the proxy-mode, hence not applying the timeout.

Yes, but after I fixed it, tests are still failing. See https://pastebin.com/F6eLN5hH (posted in https://github.com/kubernetes/kubernetes/issues/94337#issuecomment-683467140).

**Comment by aojea** [comment]
upss, sorry, somehow I missed it.

You should check why ipvs is not honoring the timeout then, maybe your kernel version is too old? 
(you don't need kubernetes for testing this, but I'm not an ipvs expert though)

**Comment by invidian** [comment]
Hmm, I was running on Flatcar Linux stable with kernel `4.19.143`. I'll try with latest beta with `5.4.62`, maybe that helps. Thanks for suggestion!

**Comment by invidian** [comment]
I've tried on kernel 5.4.62-flatcar now and same thing is happening. Here is full log: https://pastebin.com/HZfJBtwm

**Comment by athenabot** [comment]
@aojea
If this issue has been triaged, please comment `/remove-triage unresolved`.

If you aren't able to handle this issue, consider unassigning yourself and/or adding the `help-wanted` label.

🤖 I am a bot run by vllry. 👩‍🔬

**Comment by aojea** [comment]
@invidian please check this https://github.com/kubernetes/kubernetes/pull/94968#issuecomment-713283506, I think that you are hitting the same problem

**Comment by invidian** [comment]
>@invidian please check this #94968 (comment), I think that you are hitting the same problem

Thanks for the ping. I'll check it!

**Comment by thockin** [comment]
What should we do with this issue?

**Comment by invidian** [comment]
@thockin sorry about that, I didn't get time to test that. I was hoping to get sonobuoy updated, so I can just run my regular tests. Trying to run this single test fixed in #94968 right now..

**Comment by invidian** [comment]
Okay, I figured out how to run single conformance test finally (it wasn't straightforward :smile:) and it seems both tests are passing now, so :tada: So, if we want to keep this issue focused, I think my main concern is addressed (which was failing timeout in test) and I have a workaround for detecting proxy mode in e2e tests, so we can close the issue.

Going back to:
>I assume that is the reason why the detector has a check for localhost hardcoded, and looking at your scripts seems that you are binding it directly to the host ip. I don't have any strong opinion on this, but seems you've found a workaround, another option is to bind kube-proxy to 0.0.0.0 so it listens in both localhost and the host ip.

It seems listening either on `localhost` or `0.0.0.0` is sort of all-or-nothing solution, it would be nice if e.g. there was a way to listen on multiple IP addressed or something.

Another thing I found while testing `should have session affinity timeout work for NodePort service` is, the following kube-proxy config is not conformant either:
```yaml
nodePortAddresses:
  - 127.0.0.0/8
```

I had this config, as I do not use nodePort on my cluster, so I disabled that. Adding nodes CIDR to this list of course helps. Perhaps someone finds that useful.

In any case, thanks a lot for your help @aojea, I'll close the issue then.
